### PR TITLE
NuGetV3 API support

### PR DESCRIPTION
**Pull request by gpeshev1**

### CODE REVIEW

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 1]
```diff
@@ -33,26 +33,54 @@ public async Task<NuGetPackage> GetPackageWithFullDependencyTree(string id, stri
...
-            XElement propertiesElement = nuGetPackageXmlDoc
-                .Element(this.xmlns + Constants.EntryElem)
-                .Element(this.xmlnsm + Constants.PropertiesElem);
-
+            IList<NuGetPackage> dependencies = new List<NuGetPackage>();
             NuGetPackage nuGetPackage = new NuGetPackage();
-            nuGetPackage.Id = nuGetPackageXmlDoc
-                .Element(this.xmlns + Constants.EntryElem)
-                .Element(this.xmlns + Constants.TitleElem).Value;
+            if (this.isV2)
```
Comment (RuzmanovDev) [inline Sitefinity CLI/PackageManagement/NuGetApiClient.cs:45]: Encapsulate the logic for v3 and v2 into separate methods for better readability.

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 2]
```diff
@@ -33,26 +33,54 @@ public async Task<NuGetPackage> GetPackageWithFullDependencyTree(string id, stri
...
-            string dependenciesString = propertiesElement.Element(this.xmlnsd + Constants.DependenciesElem).Value;
+                string dependenciesString = propertiesElement.Element(this.xmlnsd + Constants.DependenciesElem).Value;
+
+                dependencies = this.ParseDependencies(dependenciesString);
+            }
+            else
+            {
+                var packageNamespace = nuGetPackageXmlDoc.Root.GetDefaultNamespace();
+                var elementPackage = nuGetPackageXmlDoc.Element(packageNamespace + Constants.PackageElem);
+                var metadataNamespace = elementPackage.Descendants().First().GetDefaultNamespace();
```
Comment (RuzmanovDev) [inline Sitefinity CLI/PackageManagement/NuGetApiClient.cs:65]: are we sure that the meta is the first decendant? Is there a way of checking it?

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 3]
```diff
@@ -33,26 +33,54 @@ public async Task<NuGetPackage> GetPackageWithFullDependencyTree(string id, stri
...
+                var elementsMetadata = elementPackage.Element(metadataNamespace + Constants.MetadataElem);
+                var groupElementsDependencies = elementsMetadata.Element(metadataNamespace + Constants.DependenciesEl);
 
-            IList<NuGetPackage> dependencies = this.ParseDependencies(dependenciesString);
+                if (groupElementsDependencies != null)
+                {
+                    var v3Regex = new Regex("^.[A-Z]{4}[a-z]{8}\\d.\\d\\b[.0-9]*$", RegexOptions.Compiled);
```
Comment (RuzmanovDev) [inline Sitefinity CLI/PackageManagement/NuGetApiClient.cs:71]: It is not clear what v3regex is? Think of a better name

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 4]
```diff
@@ -135,16 +163,34 @@ private async Task<XDocument> GetPackageXmlDocument(string id, string version, I
             }
 
             HttpResponseMessage response = null;
-            foreach (string source in sources)
+            var v3Sources = sources.Where(x => x.Contains(Constants.ApiV3SourceString));
+            var v2Sources = sources.Except(v3Sources);
+            foreach (string source in v3Sources)
```
Comment (RuzmanovDev) [inline Sitefinity CLI/PackageManagement/NuGetApiClient.cs:168]: extract into a method for better readability.

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 5]
```diff
@@ -135,16 +163,34 @@ private async Task<XDocument> GetPackageXmlDocument(string id, string version, I
...
+                response = await httpClient.GetAsync($"{sourceUrl}/{loweredId}/{version}/{loweredId}.nuspec");
                 if (response.StatusCode == HttpStatusCode.OK)
                 {
+                    this.isV2 = false;
                     break;
                 }
             }
 
+            if (isV2)
+            {
```
Comment (RuzmanovDev) [inline Sitefinity CLI/PackageManagement/NuGetApiClient.cs:181]: extract into a method for better readability.

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 6]
```diff
@@ -249,6 +295,44 @@ private IList<NuGetPackage> ParseDependencies(string dependenciesString)
             return dependencies;
         }
 
+        private async Task<string> GetBaseAddress(HttpClient httpClient, string source)
```
Comment (RuzmanovDev) [inline Sitefinity CLI/PackageManagement/NuGetApiClient.cs:298]: What is this base address?

**Review by RuzmanovDev — CHANGES_REQUESTED** [review]

File: Sitefinity CLI/PackageManagement/SitefinityPackageManager.cs [hunk 7]
```diff
@@ -627,6 +627,8 @@ public void SetTargetFramework(IEnumerable<string> sitefinityProjectFilePaths, s
 
         private const string PublicNuGetSource = "https://nuget.org/api/v2/";
```
[the remark this answers is not in the document]
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/SitefinityPackageManager.cs:628]: we can

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 8]
```diff
@@ -261,7 +345,9 @@ private IList<NuGetPackage> ParseDependencies(string dependenciesString)
 
         private readonly IDictionary<string, XDocument> nuGetPackageXmlDocumentCache;
 
+        private bool isV2;
```
[the remark this answers is not in the document]
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:348]: That's true, unfortunately the method is asynchronous and we can't out anything.

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs (hunk 2 above)
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:65]: From what I found, it is a convention and all nuspec index files should have the structure - <package> -> <metadata> -> <dependencies> -> <dependency>

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs (hunk 1 above)
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:45]: done

File: Sitefinity CLI/PackageManagement/SitefinityPackageManager.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/SitefinityPackageManager.cs:628]: removed

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:348]: the logic is changed now

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs (hunk 3 above)
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:71]: the regex is removed

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs (hunk 4 above)
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:168]: the code is refactored

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs (hunk 5 above)
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:181]: the code is refactored

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs (hunk 6 above)
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs:298]: We fetch the base URL from the service index, as it is recommended, I have added a comment to clarify this a bit

File: Sitefinity CLI/PackageManagement/NuGetApiClient.cs [hunk 9]
```diff
@@ -33,26 +35,22 @@ public async Task<NuGetPackage> GetPackageWithFullDependencyTree(string id, stri
...
-                .Element(this.xmlns + Constants.TitleElem).Value;
-
-            nuGetPackage.Version = propertiesElement.Element(this.xmlnsd + Constants.VersionElem).Value;
-
-            string dependenciesString = propertiesElement.Element(this.xmlnsd + Constants.DependenciesElem).Value;
-
-            IList<NuGetPackage> dependencies = this.ParseDependencies(dependenciesString);
+            if (nuGetPackageXmlDoc.ProtoVersion == ProtocolVersion.NuGetAPIV2)
+            {
+                dependencies = new List<NuGetPackage>(this.ParseDependenciesV2(nuGetPackageXmlDoc, nuGetPackage));
```
[the remark this answers is not in the document]
Comment (gpeshev1) [reply Sitefinity CLI/PackageManagement/NuGetApiClient.cs]: fixed

**Review by RuzmanovDev — APPROVED** [review]

**Review by bonny-bonev — APPROVED** [review]

**Pull request merged**